File: src/main/java/dev/psygamer/wireframe/Wireframe.java
package dev.psygamer.wireframe;

import dev.psygamer.wireframe.event.EventBus;
import dev.psygamer.wireframe.event.EventBusRegistrator;
import dev.psygamer.wireframe.event.api.IEventBus;
import dev.psygamer.wireframe.test.item_display.ItemDisplayBlock;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;

/** The main class of the Wireframe Framework */
public class Wireframe {
	
	public static final String MODID = "wireframe";
	public static final Logger LOGGER = LogManager.getLogger("Wireframe");
	public static final IEventBus EVENT_BUS = new EventBus();
	
	/** All registered mods which have Wireframe as dependency. */
	private static final List<Mod> mods = Lists.newArrayList(new InternalMod());
	
	/** Internal, do not use! */
	private static Internal internal;
	
	public static void register(final Mod mod) {
		mods.add(mod);
	}
	
	/** Returns a list of all child mods. */
	public static ImmutableList<Mod> getMods() {
		return ImmutableList.copyOf(mods);
	}
	
	/** Internal, do not use! */
	public static Internal getInternal() {
		return internal;
	}
	
	/**
	 * A class which all child mods must extend.
	 * This contains basic information about a child mod.
	 */
	public static class Mod {
		
		private final String modID, modName, modVersion, rootPackage;
		private final EventBus modEventBus = new EventBus();
		
		private final net.minecraftforge.eventbus.api.IEventBus internalModEventBus;
		
		/** Constructs a new child mod container */
		protected Mod(final String modID, final String modName, final String modVersion) {
			this.modID = modID;
			this.modName = modName;
			this.modVersion = modVersion;
			
			this.rootPackage = this.getClass()
								   .getPackage()
								   .getName();
			
			this.internalModEventBus = net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext.get()
																								 .getModEventBus();
		}
		
		public String getModID() {
			return this.modID;
		}
		
		public String getModName() {
			return this.modName;
		}
		
		public String getModVersion() {
			return this.modVersion;
		}
		
		/** The package-path of class which extends this class. */
		public String getRootPackage() {
			return this.rootPackage;
		}
		
		public EventBus getModEventBus() {
			return this.modEventBus;
		}
		
		/** Internal, do not use! */
		public net.minecraftforge.eventbus.api.IEventBus getInternalModEventBus() {
			return this.internalModEventBus;
		}
	}
	
	/** Internal, do not use! */
	@net.minecraftforge.fml.common.Mod(Wireframe.MODID)
	public static final class Internal {
		
		public Internal() {
			Wireframe.internal = this;
			
			final net.minecraftforge.eventbus.api.IEventBus modEventBus =
					net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext.get()
																			  .getModEventBus();
			
			modEventBus.addListener(this::onModConstruct);
			
			new ItemDisplayBlock();
		}
		
		private void onModConstruct(final net.minecraftforge.fml.event.lifecycle.FMLConstructModEvent event) {
			EventBusRegistrator.registerModLoadingEventBuses();
			EventBusRegistrator.registerWireframeEventBuses();
		}
	}
	
	/** Internal, do not use! */
	private static final class InternalMod extends Mod {
		
		private InternalMod() {
			super("wireframe", "Wireframe", "1.0");
		}
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/Wireframe.kt
package dev.psygamer.wireframe

import org.apache.logging.log4j.*
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import dev.psygamer.wireframe.api.registry.PacketRegistry
import dev.psygamer.wireframe.debug.WireframeDebugger
import dev.psygamer.wireframe.debug.logging.*
import dev.psygamer.wireframe.event.*
import dev.psygamer.wireframe.event.api.IEventBus
import dev.psygamer.wireframe.test.*

object Wireframe {

	const val NAME = "Wireframe"
	const val MODID = "wireframe"
	const val VERSION = "1.0"

	@JvmField
	val LOGGER: Logger = LogManager.getLogger(NAME)

	@JvmField
	val CHAT_LOGGER: SimpleLogger =
		if (WireframeDebugger.ENABLED)
			ChatLogger(name = NAME, consoleLogger = LOGGER)
		else // We only want to use the ChatLogger in a debug environment.
			VoidLogger

	@JvmField
	val EVENT_BUS: IEventBus = EventBus()

	@JvmStatic
	val mods: List<Mod> = mutableListOf()

	@JvmStatic
	var nativeModImplementation: NativeMod? = null
		private set

	open class Mod(val modID: String, val modName: String, val modVersion: String) {

		val rootPackage: String = this.javaClass.`package`.name

		val eventBus: IEventBus = EventBus()
		val nativeEventBus: net.minecraftforge.eventbus.api.IEventBus = MOD_BUS

		init {
			(mods as MutableList).add(this)
		}
	}

	private object WireframeMod : Mod(MODID, NAME, VERSION)

	@net.minecraftforge.fml.common.Mod(MODID)
	object NativeMod {

		init {
			nativeModImplementation = this

			/* Force-init the internal implementation.
			 * This will probably cause issues with child mods which don't have this way of force-init-ing their extension class.
			 *
			 * TODO Do this automatically
			 */ WireframeMod

			BlockTest()
			EventBusRegistrator.register()
			PacketRegistry.register(TestPacket::class.java, TestPacketDecoder)
		}
	}
}